internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -140,27 +142,55 @@ func (r *AgentServiceConfigReconciler) Reconcile(origCtx context.Context, req ct
 		return reconcile.Result{}, nil
 	}
 
-	for _, f := range []func(context.Context, logrus.FieldLogger, *aiv1beta1.AgentServiceConfig) error{
-		r.ensureFilesystemStorage,
-		r.ensureDatabaseStorage,
-		r.ensureAgentService,
-		r.ensureServiceMonitor,
-		r.ensureAgentRoute,
-		r.ensureAgentLocalAuthSecret,
-		r.ensurePostgresSecret,
-		r.ensureIngressCertCM,
-		r.ensureAssistedCM,
-		r.ensureAssistedServiceDeployment,
+	for _, component := range []struct {
+		name   string
+		reason string
+		fn     NewComponentFn
+	}{
+		{"FilesystemStorage", aiv1beta1.ReasonStorageFailure, r.newFilesystemPVC},
+		{"DatabaseStorage", aiv1beta1.ReasonStorageFailure, r.newDatabasePVC},
+		{"AgentService", aiv1beta1.ReasonAgentServiceFailure, r.newAgentService},
+		{"ServiceMonitor", aiv1beta1.ReasonAgentServiceMonitorFailure, r.newServiceMonitor},
+		{"AgentRoute", aiv1beta1.ReasonAgentRouteFailure, r.newAgentRoute},
+		{"AgentLocalAuthSecret", aiv1beta1.ReasonAgentLocalAuthSecretFailure, r.newAgentLocalAuthSecret},
+		{"DatabaseSecret", aiv1beta1.ReasonPostgresSecretFailure, r.newPostgresSecret},
+		{"IngressCertConfigMap", aiv1beta1.ReasonIngressCertFailure, r.newIngressCertCM},
+		{"AssistedServiceConfigMap", aiv1beta1.ReasonConfigFailure, r.newAssistedCM},
+		{"AssistedServiceDeployment", aiv1beta1.ReasonDeploymentFailure, r.newAssistedServiceDeployment},
 	} {
-		err := f(ctx, log, instance)
+		obj, mutateFn, err := component.fn(ctx, log, instance)
 		if err != nil {
-			log.Error(err, "Failed reconcile")
+			msg := "Failed to generate definition for " + component.name + ": " + err.Error()
+			log.Error(err, msg)
+			conditionsv1.SetStatusConditionNoHeartbeat(&instance.Status.Conditions, conditionsv1.Condition{
+				Type:    aiv1beta1.ConditionReconcileCompleted,
+				Status:  corev1.ConditionFalse,
+				Reason:  aiv1beta1.ReasonStorageFailure,
```

Comment:
This seems like a strange choice for all of them.
What are the differences between these reasons and why does a failure here indicate a storage failure?

Reply:
That's a bug. Fixed.